### PR TITLE
PTX-2759 Add PX Security validation and basic regression test

### PR DESCRIPTION
**Pull request by nikolaypopov**
Add PX Security validation and basic P0 regression tests

Signed-off-by: nikolaypopov <nikolay.popov86@gmail.com>

### CODE REVIEW

**Review by piyush-nimbalkar — APPROVED** [review]
lgtm after minor comments

File: pkg/util/test/util.go [hunk 1]
```diff
@@ -1726,6 +1744,212 @@ func validateImageTag(tag, namespace string, listOptions map[string]string) erro
...
+		cluster.Spec.Security.Enabled {
+		logrus.Infof("PX Security is enabled")
+		return ValidateSecurityEnabled(cluster, timeout, interval)
+	}
+
+	logrus.Infof("PX Security is not enabled")
+	return ValidateSecurityDisabled(cluster, previouslyEnabled, timeout, interval)
+}
+
+// ValidateSecurityEnabled validates PX Security components are enabled`/running as expected
```
[the remark this answers is not in the document]
Comment (nikolaypopov) [reply pkg/util/test/util.go:1759]: Done

File: pkg/util/test/util.go [hunk 2]
```diff
@@ -1726,6 +1744,212 @@ func validateImageTag(tag, namespace string, listOptions map[string]string) erro
...
+			}
+
+			// Validate Security ENv vars in Stork pods
+			if err := validateStorkSecurityEnvVar(cluster, storkDp, timeout, interval); err != nil {
+				return "", true, fmt.Errorf("failed to validate Stork Security ENV vars, err %v", err)
+			}
+		}
+
+		if _, err := coreops.Instance().GetSecret("px-admin-token", cluster.Namespace); err != nil {
+			return "", true, fmt.Errorf("failed to find secret secret px-admin-token, err %v", err)
```
[the remark this answers is not in the document]
Comment (nikolaypopov) [reply pkg/util/test/util.go:1780]: Done

File: pkg/util/test/util.go [hunk 3]
```diff
@@ -1726,6 +1744,212 @@ func validateImageTag(tag, namespace string, listOptions map[string]string) erro
...
+
+			// Validate Security ENv vars in Stork pods
+			if err := validateStorkSecurityEnvVar(cluster, storkDp, timeout, interval); err != nil {
+				return "", true, fmt.Errorf("failed to validate Stork Security ENV vars, err %v", err)
+			}
+		}
+
+		if previouslyEnabled {
+			_, err := coreops.Instance().GetSecret("px-admin-token", cluster.Namespace)
+			if !errors.IsNotFound(err) {
```
[the remark this answers is not in the document]
Comment (nikolaypopov) [reply pkg/util/test/util.go:1827]: There are 4 secrets that Security enabled creates:
px-admin-token
px-user-token
px-shared-secret
px-system-secrets
Now the *-token secrets actually get deleted when you disable Security, only *-secrets remain present. I don't know if this is expected or not, but that is the behavior right now.

File: pkg/util/test/util.go [hunk 4]
```diff
@@ -1726,6 +1744,212 @@ func validateImageTag(tag, namespace string, listOptions map[string]string) erro
...
+			}
+
+			if securityEnabled {
+				logrus.Debugf("Value for %s env var in Stork pod [%s]: expected: %v, actual: %v", StorkPxJwtIssuerEnvVarName, pod.Name, DefaultStorkPxJwtIssuerEnvVarValue, pxJwtIssuerEnvVar)
+				logrus.Debugf("Value for %s env var in Stork pod [%s]: expected: %v, actual: %v", StorkPxSharedSecretEnvVarName, pod.Name, DefaultStorkPxSharedSecretEnvVarValue, pxSharedSecretEnvVar)
+			}
+			numberOfPods++
+		}
+
+		// TODO: Hardcoding this to 3 instead of len(pods), because the previous ValidateDeloyment() step might have not validated the updated deloyment
```
[the remark this answers is not in the document]
Comment (nikolaypopov) [reply pkg/util/test/util.go:1939]: Done

File: test/integration_test/basic_test.go [hunk 5]
```diff
@@ -789,6 +797,58 @@ func AlertManagerRegression(tc *types.TestCase) func(*testing.T) {
...
+		// Enable Security and validate
+		logrus.Info("Enable Security and validate")
+		var security *corev1.SecuritySpec
+		updateParamFunc := func(cluster *corev1.StorageCluster) *corev1.StorageCluster {
+			if cluster.Spec.Security == nil {
+				security = &corev1.SecuritySpec{
+					Enabled: true,
+				}
+				cluster.Spec.Security = security
+			}
```
[the remark this answers is not in the document]
Comment (nikolaypopov) [reply test/integration_test/basic_test.go:832]: Will change it to
`            if cluster.Spec.Security == nil {
                cluster.Spec.Security = &corev1.SecuritySpec{}
            }
            cluster.Spec.Security.Enabled = true`
This way we can get rid of `security` var all together

File: pkg/util/test/util.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (piyush-nimbalkar) [reply pkg/util/test/util.go:1827]: Got it. That is what I was saying on slack. We don't necessarily need the token if security is disabled. Also, we anyway recreate the tokens when security is enabled again. What we need to keep are the system and shared secrets.

Can we do different checks for the tokens and secrets in that case? Expect the tokens to be deleted even if previously enabled. While for the secrets they will exist if previously enabled.

[the remark this answers is not in the document]
Comment (piyush-nimbalkar) [reply pkg/util/test/util.go:1827]: For your reference - https://github.com/libopenstorage/operator/blob/master/drivers/storage/portworx/component/auth.go#L137

File: test/integration_test/basic_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (piyush-nimbalkar) [reply test/integration_test/basic_test.go:832]: Looks good!

File: pkg/util/test/util.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (nikolaypopov) [reply pkg/util/test/util.go]: Done

**Pull request merged**